paasta_tools/cleanup_kubernetes_jobs.py

```diff
@@ -95,44 +104,118 @@ def alert_state_change(application: Application, soa_dir: str) -> Generator:
+            )
+            raise StatefulSetsAreNotSupportedError
+    return None
+
+
+def get_applications_to_kill(
+    applications_dict: Dict[Tuple[str, str], List[Application]],
+    cluster: str,
+    valid_services: Set[Tuple[str, str]],
+    kube_client: KubeClient,
+    soa_dir: str,
+) -> List[Application]:
+    """
+
+    :param applications_dict: A dictionary with (service, instance) as keys and a list of applications for each tuple
+    :param cluster: paasta cluster
+    :param valid_services: a set with the valid (service, instance) tuples for this cluster
+    :param kube_client:
+    :param soa_dir: The SOA config directory to read from
+    :return: list of applications to kill
+    """
+    log.info("Determining apps to be killed")
+
+    applications_to_kill: List[Application] = []
+    for instance_tuple, applications in applications_dict.items():
+        service = instance_tuple[0]
+        instance = instance_tuple[1]
+        instance_config = load_kubernetes_service_config(
+            cluster=cluster, service=service, instance=instance, soa_dir=soa_dir
+        )
+        if len(applications) == 1:
+            application = applications[0]
+            if (
+                (service, instance) not in valid_services
+                or application.kube_deployment.namespace
+                != instance_config.get_namespace()
+            ):
+                applications_to_kill.append(application)
+        elif len(applications) > 1:
+            not_bouncing = instance_is_not_bouncing(
+                instance_config, applications, kube_client
+            )
+            for application in applications:
+                if (service, instance) not in valid_services or (
+                    application.kube_deployment.namespace
+                    != instance_config.get_namespace()
+                    and not_bouncing
+                ):
+                    applications_to_kill.append(application)
+    return applications_to_kill
+
+
 def cleanup_unused_apps(
-    soa_dir: str, kill_threshold: float = 0.5, force: bool = False
+    soa_dir: str,
+    cluster: str,
+    kill_threshold: float = 0.5,
+    force: bool = False,
 ) -> None:
     """Clean up old or invalid jobs/apps from kubernetes. Retrieves
     both a list of apps currently in kubernetes and a list of valid
     app ids in order to determine what to kill.
 
     :param soa_dir: The SOA config directory to read from
+    :param cluster: paasta cluster to clean
     :param kill_threshold: The decimal fraction of apps we think is
         sane to kill when this job runs.
     :param force: Force the cleanup if we are above the kill_threshold"""
     log.info("Creating KubeClient")
     kube_client = KubeClient()
 
     log.info("Loading running Kubernetes apps")
-    applications = list_namespaced_applications(
-        kube_client, "paasta", APPLICATION_TYPES
-    )
-
+    applications_dict = list_all_applications(kube_client, APPLICATION_TYPES)
```

Comment:
could this still be called `applications`? future-us will thank us if this gets refactored to not return a dict :)

(and mypy and co will save us from any mistakes where we treat this as anything but a dict without the variable name referencing the type)